File: components/support/nimbus-fml/fixtures/android/runtime/MockNimbus.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@file:Suppress("InvalidPackageDeclaration")

package org.mozilla.experiments.nimbus

import android.content.Context
import org.json.JSONObject

class MockNimbus(override val context: Context, val map: Map<String, JSONObject>) : FeaturesInterface {

    constructor(vararg pairs: Pair<String, String>, context: Context = Context()) : this(
        context,
        pairs.toMap().mapValues { entry ->
            JSONObject(entry.value)
        },
    )

    init {
        NullVariables.instance.setContext(context)
    }

    override fun getVariables(featureId: String, recordExposureEvent: Boolean): Variables =
        map[featureId]?.let { json -> JSONVariables(context, json) } ?: NullVariables.instance

    private var exposureCounts = mutableMapOf<String, Int>()

    override fun recordExposureEvent(featureId: String, experimentSlug: String?) {
        if (map[featureId] != null) {
            exposureCounts[featureId] = getExposureCount(featureId) + 1
        }
    }

    fun getExposureCount(featureId: String) = exposureCounts.getOrDefault(featureId, 0)

    fun isExposed(featureId: String) = getExposureCount(featureId) > 0
}

File: components/support/nimbus-fml/fixtures/ios/runtime/MockNimbus.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import Foundation

public class MockNimbus: FeaturesInterface {
    var map: [String: Any] = [:]
    public init(_ pairs: (String, String)...) {
        for (key, value) in pairs {
            if let data = value.data(using: .utf8) {
                do {
                    let json = try JSONSerialization.jsonObject(with: data, options: .mutableContainers)
                    map[key] = json
                } catch {
                    fatalError("Invalid JSON Passed")
                }
            }
        }
    }

    public func getVariables(featureId: String, recordExposureEvent: Bool = true) -> Variables? {
        if let json = map[featureId] {
            return JSONVariables(with: json as! [String: Any])
        }
        return nil
    }

    private var exposureCounts: [String: Int] = [:]

    public func recordExposureEvent(featureId: String) {
        if map[featureId] != nil {
            exposureCounts[featureId] = getExposureCount(featureId: featureId) + 1
        }
    }

    public func getExposureCount(featureId: String) -> Int {
        if let count = exposureCounts[featureId] {
            return count
        }
        return 0
    }

    public func isExposed(featureId: String) -> Bool {
        getExposureCount(featureId: featureId) > 0
    }
}
